message_subscribe_ui/src/Plugin/Block/Subscriptions.php

```diff
@@ -0,0 +1,223 @@
+   *   The current route match service.
+   * @param \Drupal\Core\Session\AccountProxyInterface $current_user
+   *   The current user.
+   * @param \Drupal\flag\FlagServiceInterface $flag_service
+   *   The flag service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, FormBuilderInterface $form_builder, SubscribersInterface $subscribers, RouteMatchInterface $route_match, AccountProxyInterface $current_user, FlagServiceInterface $flag_service) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->currentUser = $current_user;
+    $this->flagService = $flag_service;
+    $this->formBuilder = $form_builder;
+    $this->routeMatch = $route_match;
+    $this->subscribers = $subscribers;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('form_builder'),
+      $container->get('message_subscribe.subscribers'),
+      $container->get('current_route_match'),
+      $container->get('current_user'),
+      $container->get('flag')
+    );
+  }
+
+  /**
+   * Helper method to retrieve the current page entity.
+   *
+   * @return \Drupal\Core\Entity\EntityInterface|null
+   *   The entity for the current route.
+   */
+  protected function getCurrentEntity() {
+    // Let's look up in the route object for the name of upcasted values.
+    foreach ($this->routeMatch->getParameters() as $parameter) {
+      if ($parameter instanceof EntityInterface) {
+        return $parameter;
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build() {
+    return $this->formBuilder->getForm($this);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'message_subscribe_ui_block';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $form = [];
+    $subscribable = FALSE;
+    if ($entity = $this->getCurrentEntity()) {
+      // Add the entity being viewed.
+      $entities = [$entity];
+      $entities += $entity->referencedEntities();
+      if (!empty($entities)) {
+        $form['description'] = [
+          '#type' => 'markup',
+          '#markup' => $this->t('Manage all <a href=":url">subscriptions</a>.', [':url' => Url::fromRoute('message_subscribe_ui.tab', ['user' => \Drupal::currentUser()->id()])->toString()]),
+        ];
+        $form['subscriptions'] = [
+          '#type' => 'container',
+          '#tree' => TRUE,
+        ];
+        foreach ($entities as $referenced_entity) {
+          $flags = $this->subscribers->getFlags($referenced_entity->getEntityTypeId(), $referenced_entity->bundle(), $this->currentUser);
+          if (!empty($flags)) {
+            // @todo Support multiple subscription flags per-entity if there is
```

Comment:
Well, the `getFlags` method returns an array, and here I just grab one as I assume there won't be more than one subscribe flag per entity type. However, there *could* be in theory anyway, thus the `@todo`. I can replace the `@todo` with a simpler note explaining this logic...